### DISCUSSION TITLE
Chunked upload / partial updates to object store

### DISCUSSION THREAD

**Discussion by matthiasharrer** [nats-io/nats-server#4689] [Q&A] [answered]
Hi,

I was just looking through the object store interfaces, evaluating it for my use cases.
One thing I am missing is a partial update of an object. A bit like multipart upload to S3 and the like.
As my understanding is, that objects are basically just streams this should technically not be an issue right?

For my use case, which I am not sure if it's an appropriate use case for object store in general:

Users are uploading large files (100 MBs to multiple GBs) using https://tus.io/ for increasing upload reliability, where users upload the file in chunks and in case of any failure can restart the upload where it was left off. 
Failures might be the users internet connection, but also for example server restarts / upgrades.

So I think the basic requirement for using NATS as a store would be to add new chunks at the end of a file and to somehow mark a file as incomplete / complete (which could be done as metadata I guess)

Does such functionality already exist?
Is this a proper use case for the object store?

**Comment by aricart** [comment, marked as answer]
this actually doesn't work with ObjectStore - the reason being that each chunk is considered to be in-order in the stream. So it is not possible to just update chunk #123 and have it be at the right offset. With that said, there could be different strategies that could be used to implement this with a little more work from the client. You could think of the asset as being made of 100 logical entries on your side (regardless what the storage chunks are), and you could update each of these entries as you see fit. ObjectStore will ensure that the chunks are correct, but you could simply replace the entry that failed as a whole. - IE store `large.1, large.2..., large.N` - your code can then simply request each of these objects and then put it back together yourself.

**Reply by matthiasharrer** [reply]
Hey thanks for the quick answer!

Actually for me it would be totally ok if I could just add new chunks at the end of the file / stream (I think .. )
But I guess that's also not curretnly possible with the ObjectStore.
Maybe I will continue thinking about a more bespoke solution as you mentioned.
